// yoastseo/js/helpers/getFunctionWords.js
"use strict";
/*
 * The script collects all the lists of function words per language and returns this collection to a Researcher or a
 * stringProcessing script
 */

var germanFunctionWords = require("../researches/german/functionWords.js")();
var englishFunctionWords = require("../researches/english/functionWords.js")();
var dutchFunctionWords = require("../researches/dutch/functionWords.js")();
var spanishFunctionWords = require("../researches/spanish/functionWords.js")();
var italianFunctionWords = require("../researches/italian/functionWords.js")();
var frenchFunctionWords = require("../researches/french/functionWords.js")();
var portugueseFunctionWords = require("../researches/portuguese/functionWords.js")();
var russianFunctionWords = require("../researches/russian/functionWords.js")();
module.exports = function () {
    return {
        en: englishFunctionWords,
        de: germanFunctionWords,
        nl: dutchFunctionWords,
        fr: frenchFunctionWords,
        es: spanishFunctionWords,
        it: italianFunctionWords,
        pt: portugueseFunctionWords,
        ru: russianFunctionWords
    };
};
